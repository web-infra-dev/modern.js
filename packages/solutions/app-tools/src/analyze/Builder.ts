import type { BuildOptions, BuildResult } from 'esbuild';
import { build } from 'esbuild';

export class Builder {
  private buildRes: BuildResult | undefined;

  async build(options: BuildOptions) {
    this.buildRes = await build(options);
  }

  stop() {
    this.buildRes?.stop && this.buildRes?.stop();
  }
}

export const loaderBuilder = new Builder();
export const serverLoaderBuilder = new Builder();
