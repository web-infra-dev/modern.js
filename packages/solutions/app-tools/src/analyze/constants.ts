export const JS_EXTENSIONS = ['.js', '.ts', '.jsx', '.tsx'];

export const INDEX_FILE_NAME = 'index';

export const APP_FILE_NAME = 'App';

export const PAGES_DIR_NAME = 'pages';

export const NESTED_ROUTES_DIR = 'routes';

export const FILE_SYSTEM_ROUTES_FILE_NAME = 'routes.js';

export const LOADER_EXPORT_NAME = 'loader';

export const TEMP_LOADERS_DIR = '__loaders__';

export const ENTRY_POINT_FILE_NAME = 'index.js';

export const ENTRY_BOOTSTRAP_FILE_NAME = 'bootstrap.js';

export const FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP = /^\[(\S+)\]([*+?]?)$/;

export const FILE_SYSTEM_ROUTES_LAYOUT = '_layout';

export const FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT = '_app';

export const FILE_SYSTEM_ROUTES_INDEX = 'index';

export const FILE_SYSTEM_ROUTES_IGNORED_REGEX =
  /\.(d|test|spec|e2e)\.(js|jsx|ts|tsx)$/;

export const HTML_PARTIALS_FOLDER = 'html';

export const HTML_PARTIALS_EXTENSIONS = ['.htm', '.html', '.ejs'];

export const FILE_SYSTEM_ROUTES_COMPONENTS_DIR = 'internal_components';

export const NESTED_ROUTE = {
  LAYOUT_FILE: 'layout',
  LAYOUT_LOADER_FILE: 'layout.loader',
  PAGE_FILE: 'page',
  PAGE_LOADER_FILE: 'page.loader',
  LOADING_FILE: 'loading',
  ERROR_FILE: 'error',
  LOADER_FILE: 'loader',
};
