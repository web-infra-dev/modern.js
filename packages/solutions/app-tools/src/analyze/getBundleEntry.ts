import path from 'path';
import {
  ensureAbsolutePath,
  fs,
  findExists,
  MAIN_ENTRY_NAME,
} from '@modern-js/utils';
import type { Entrypoint } from '@modern-js/types';
import type { AppNormalizedConfig, IAppContext } from '../types';
import { getFileSystemEntry } from './getFileSystemEntry';
import { JS_EXTENSIONS } from './constants';

const ensureExtensions = (file: string) => {
  if (!path.extname(file)) {
    return findExists(JS_EXTENSIONS.map(ext => `${file}${ext}`)) || file;
  }
  return file;
};

const ifAlreadyExists = (
  entrypoints: Entrypoint[],
  checked: Entrypoint,
): boolean =>
  entrypoints.some(entrypoint => {
    if (
      ensureExtensions(entrypoint.entry) === ensureExtensions(checked.entry)
    ) {
      // reset entryName
      checked.entryName = entrypoint.entryName;
      return true;
    }
    // filesystem routes entrypoint conflict with normal entrypoint.
    if (
      entrypoint.entry.startsWith(checked.entry) ||
      checked.entry.startsWith(entrypoint.entry)
    ) {
      throw new Error(
        `Entry configuration conflicts\n Your configuration: ${checked.entry}.\n Default entrypoint: ${entrypoint.entry}\n Please reset source.entries or set source.disableDefaultEntries to disable the default entry rules.`,
      );
    }

    return false;
  });

export const getBundleEntry = (
  appContext: IAppContext,
  config: AppNormalizedConfig,
) => {
  const { appDirectory, packageName } = appContext;
  const {
    source: { disableDefaultEntries, entries, entriesDir },
  } = config;

  const defaults = disableDefaultEntries
    ? []
    : getFileSystemEntry(appContext, config);

  // merge entrypoints from user config with directory convention.
  if (entries) {
    Object.keys(entries).forEach(name => {
      const value = entries[name];
      const entrypoint =
        typeof value === 'string'
          ? {
              entryName: name,
              entry: ensureAbsolutePath(appDirectory, value),
              isAutoMount: true,
              fileSystemRoutes: fs
                .statSync(ensureAbsolutePath(appDirectory, value))
                .isDirectory()
                ? {}
                : undefined,
            }
          : {
              entryName: name,
              entry: ensureAbsolutePath(appDirectory, value.entry),
              isAutoMount: !value.disableMount,
              fileSystemRoutes: value.enableFileSystemRoutes ? {} : undefined,
            };

      if (!ifAlreadyExists(defaults, entrypoint)) {
        defaults.push(entrypoint);
      }
    });
  }

  if (!disableDefaultEntries) {
    // find main entry point which server route is '/'.
    const entriesDirAbs = ensureAbsolutePath(appDirectory, entriesDir || '');
    const found = defaults.find(
      ({ entryName, entry, nestedRoutesEntry = '' }) =>
        entryName === packageName ||
        path.dirname(entry) === entriesDirAbs ||
        path.dirname(nestedRoutesEntry) === entriesDirAbs,
    );
    found && (found.entryName = MAIN_ENTRY_NAME);
  }

  return defaults;
};
