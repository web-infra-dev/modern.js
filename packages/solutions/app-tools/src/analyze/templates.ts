import path from 'path';
import type {
  Entrypoint,
  NestedRoute,
  PageRoute,
  Route,
  RouteLegacy,
} from '@modern-js/types';
import { fs, slash } from '@modern-js/utils';
import type { RuntimePlugin } from '../types';
import { APP_CONFIG_NAME, TEMP_LOADERS_DIR } from './constants';

export const index = ({
  mountId,
  imports,
  renderFunction,
  exportStatement,
}: {
  mountId: string;
  imports: string;
  exportStatement: string;
  renderFunction: string;
}) => `
const IS_BROWSER = typeof window !== 'undefined' && window.name !== 'nodejs';
const IS_REACT18 = process.env.IS_REACT18 === 'true';
const MOUNT_ID = '${mountId}';

${imports}

let AppWrapper = null;

let root = null;

function render() {
  ${renderFunction}
}

AppWrapper = render();

${exportStatement};
`;

export const renderFunction = ({
  plugins,
  customBootstrap,
  fileSystemRoutes,
}: {
  plugins: RuntimePlugin[];
  customBootstrap?: string | false;
  fileSystemRoutes: Entrypoint['fileSystemRoutes'];
}) => {
  return `
  const finalAppConfig = {
    ...App.config,
    ...typeof ${APP_CONFIG_NAME} === 'function' ? ${APP_CONFIG_NAME}() : {},
  }

  AppWrapper = createApp({
    plugins: [
     ${plugins
       .map(
         ({ name, options, args }) =>
           `${name}({...${options}, ...finalAppConfig?.${args || name}}),`,
       )
       .join('\n')}
    ]
  })(${fileSystemRoutes ? '' : `App`})


  if(!AppWrapper.init && typeof appInit !== 'undefined') {
    AppWrapper.init = appInit;
  }


  if (IS_BROWSER) {
    ${
      customBootstrap
        ? `customBootstrap(AppWrapper);`
        : `bootstrap(AppWrapper, MOUNT_ID, root, ReactDOM);`
    }
  }

  return AppWrapper
`;
};

export const html = (partials: {
  top: string[];
  head: string[];
  body: string[];
}) => `
<!DOCTYPE html>
<html>
<head>
  <%= meta %>
  <title><%= title %></title>

  ${partials.top.join('\n')}

  <script>
    window.__assetPrefix__ = '<%= assetPrefix %>';
  </script>
  ${partials.head.join('\n')}

  <!--<?- chunksMap.css ?>-->
</head>

<body>
  <noscript>
    We're sorry but react app doesn't work properly without JavaScript enabled. Please enable it to continue.
  </noscript>
  <div id="<%= mountId %>"><!--<?- html ?>--></div>
  ${partials.body.join('\n')}
  <!--<?- chunksMap.js ?>-->
  <!--<?- SSRDataScript ?>-->
  <!--<?- bottomTemplate ?>-->
</body>

</html>
`;

export const routesForServer = ({
  routes,
}: {
  routes: (NestedRoute | PageRoute)[];
}) => {
  const loaders: string[] = [];
  const traverseRouteTree = (route: NestedRoute | PageRoute): Route => {
    let children: Route['children'];
    if ('children' in route && route.children) {
      children = route?.children?.map(traverseRouteTree);
    }
    let loader: string | undefined;

    if (route.type === 'nested') {
      if (route.loader) {
        loaders.push(route.loader);
        loader = `loader_${loaders.length - 1}`;
      }
    }

    const finalRoute = {
      ...route,
      loader,
      children,
    };
    return finalRoute;
  };

  let routesCode = `
  export const routes = [
  `;
  for (const route of routes) {
    if ('type' in route) {
      const newRoute = traverseRouteTree(route);
      routesCode += `${JSON.stringify(newRoute, null, 2).replace(
        /"(loader_[^"])"/g,
        '$1',
      )},`;
    } else {
      routesCode += `${JSON.stringify(route, null, 2)}`;
    }
  }
  routesCode += `\n];`;
  let importLoadersCode = '';
  if (loaders.length > 0) {
    importLoadersCode = loaders
      .map((loader, index) => {
        return `import loader_${index} from "${slash(loader)}"`;
      })
      .join('\n');
  }

  return `
    ${importLoadersCode}
    ${routesCode}
  `;
};

export const fileSystemRoutes = async ({
  routes,
  ssrMode,
  nestedRoutesEntry,
  entryName,
  internalDirectory,
}: {
  routes: RouteLegacy[] | (NestedRoute | PageRoute)[];
  ssrMode: 'string' | 'stream' | false;
  nestedRoutesEntry?: string;
  entryName: string;
  internalDirectory: string;
}) => {
  const loadings: string[] = [];
  const errors: string[] = [];
  const loaders: string[] = [];
  const loadersMap: Record<
    string,
    {
      routeId: string;
      filePath: string;
      inline: boolean;
    }
  > = {};

  const loadersMapFile = path.join(
    internalDirectory,
    entryName,
    TEMP_LOADERS_DIR,
    'map.json',
  );

  const importLazyCode = `
    import { lazy } from "react";
    import loadable, { lazy as loadableLazy } from "@modern-js/runtime/loadable"
  `;
  let rootLayoutCode = ``;
  let componentLoaderPath = '';
  const getDataLoaderPath = (loaderId: string) => {
    if (!ssrMode) {
      return '';
    }
    let dataLoaderPath = require.resolve(
      `@modern-js/plugin-data-loader/loader`,
    );
    if (nestedRoutesEntry) {
      dataLoaderPath = `${slash(dataLoaderPath)}?mapFile=${slash(
        loadersMapFile,
      )}&loaderId=${loaderId}!`;
    }
    return dataLoaderPath;
  };

  if (ssrMode) {
    componentLoaderPath = `${path.join(
      __dirname,
      '../builder/loaders/routerLoader',
    )}!`;
  }

  const traverseRouteTree = (route: NestedRoute | PageRoute): Route => {
    let children: Route['children'];
    if ('children' in route && route.children) {
      children = route?.children?.map(traverseRouteTree);
    }
    let loading: string | undefined;
    let error: string | undefined;
    let loader: string | undefined;
    let component = '';
    let lazyImport = null;

    if (route.type === 'nested') {
      if (route.loading) {
        loadings.push(route.loading);
        loading = `loading_${loadings.length - 1}`;
      }
      if (route.error) {
        errors.push(route.error);
        error = `error_${errors.length - 1}`;
      }
      if (route.loader) {
        loaders.push(route.loader);
        const loaderId = loaders.length - 1;
        loader = `loader_${loaderId}`;
        loadersMap[loader] = {
          routeId: route.id!,
          filePath: route.loader,
          inline: false,
        };
      }

      if (route._component) {
        if (route.isRoot) {
          rootLayoutCode = `import RootLayout from '${route._component}'`;
          component = `RootLayout`;
        } else if (ssrMode === 'string') {
          lazyImport = `() => import(/* webpackChunkName: "${route.id}" */  '${componentLoaderPath}${route._component}')`;
          component = `loadable(${lazyImport})`;
        } else {
          // csr and streaming
          lazyImport = `() => import(/* webpackChunkName: "${route.id}" */  '${componentLoaderPath}${route._component}')`;
          component = `lazy(${lazyImport})`;
        }
      }
    } else if (route._component) {
      lazyImport = `() => import('${route._component}')`;
      component = `loadable(${lazyImport})`;
    }

    const finalRoute = {
      ...route,
      lazyImport,
      loading,
      loader,
      error,
      children,
    };
    if (route._component) {
      finalRoute.component = component;
    }
    return finalRoute;
  };

  let routeComponentsCode = `
    export const routes = [
  `;
  for (const route of routes) {
    if ('type' in route) {
      const newRoute = traverseRouteTree(route);
      routeComponentsCode += `${JSON.stringify(newRoute, null, 2)
        .replace(/"(loadable.*\))"/g, '$1')
        .replace(/"(loadableLazy.*\))"/g, '$1')
        .replace(/"(\(\)[^,]+)",/g, '$1,')
        .replace(/"(lazy\(.*\))"/g, '$1')
        .replace(/"(loading_[^"])"/g, '$1')
        .replace(/"(loader_[^"])"/g, '$1')
        .replace(/"(RootLayout)"/g, '$1')
        .replace(/"(error_[^"])"/g, '$1')
        .replace(/\\"/g, '"')},`;
    } else {
      const component = `loadable(() => import('${route._component}'))`;
      const finalRoute = {
        ...route,
        component,
      };

      routeComponentsCode += `${JSON.stringify(finalRoute, null, 2)
        .replace(/"(loadable[^"]*)"/g, '$1')
        .replace(/"(lazy[^"]*)"/g, '$1')},`;
    }
  }
  routeComponentsCode += `\n];`;

  const importLoadingCode = loadings
    .map((loading, index) => {
      return `import loading_${index} from '${loading}';\n`;
    })
    .join('');

  const importErrorComponentsCode = errors
    .map((error, index) => {
      return `import error_${index} from '${error}';\n`;
    })
    .join('');

  let importLoadersCode = '';

  for (const [key, loaderInfo] of Object.entries(loadersMap)) {
    if (loaderInfo.inline) {
      importLoadersCode += `import { loader as ${key} } from "${getDataLoaderPath(
        key,
      )}${slash(loaderInfo.filePath)}";\n`;
    } else {
      importLoadersCode += `import ${key} from "${getDataLoaderPath(
        key,
      )}${slash(loaderInfo.filePath)}";\n`;
    }
  }

  await fs.ensureFile(loadersMapFile);
  await fs.writeJSON(loadersMapFile, loadersMap);

  return `
    ${importLazyCode}
    ${rootLayoutCode}
    ${importLoadingCode}
    ${importErrorComponentsCode}
    ${importLoadersCode}
    ${routeComponentsCode}
  `;
};
