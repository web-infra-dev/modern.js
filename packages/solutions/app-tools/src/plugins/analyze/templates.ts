import type { Entrypoint } from '@modern-js/types';
import type { RuntimePlugin } from '../../types';
import { APP_CONFIG_NAME } from './constants';

export const index = ({
  mountId,
  imports,
  renderFunction,
  exportStatement,
}: {
  mountId: string;
  imports: string;
  exportStatement: string;
  renderFunction: string;
}) => `
const IS_BROWSER = typeof window !== 'undefined' && window.name !== 'nodejs';
const IS_REACT18 = process.env.IS_REACT18 === 'true';
const MOUNT_ID = '${mountId}';

${imports}

let AppWrapper = null;

let root = null;

function render() {
  ${renderFunction}
}

AppWrapper = render();

${exportStatement};
`;

export const renderFunction = ({
  plugins,
  customBootstrap,
  fileSystemRoutes,
  customRuntimeConfig,
}: {
  plugins: RuntimePlugin[];
  customBootstrap?: string | false;
  customRuntimeConfig?: string | false;
  fileSystemRoutes: Entrypoint['fileSystemRoutes'];
}) => {
  const bootstrap = 'bootstrap(AppWrapper, MOUNT_ID, root, ReactDOM)';
  const runtimePlugins = `...(runtimeConfig?.plugins || []),`;

  return `
  const finalAppConfig = {
    ...App.config,
    ...typeof ${APP_CONFIG_NAME} === 'function' ? ${APP_CONFIG_NAME}() : {},
  }

  AppWrapper = createApp({
    plugins: [
     ${plugins
       .map(
         ({ name, options, args }) =>
           `${name}({...${options}, ...finalAppConfig?.${args || name}}),`,
       )
       .join('\n')}
       ${customRuntimeConfig ? runtimePlugins : ''}
    ]
  })(${fileSystemRoutes ? '' : `App`})


  if(!AppWrapper.init && typeof appInit !== 'undefined') {
    AppWrapper.init = appInit;
  }


  if (IS_BROWSER) {
    ${
      customBootstrap
        ? `customBootstrap(AppWrapper, () => ${bootstrap});`
        : `${bootstrap};`
    }
  }

  return AppWrapper
`;
};

export const html = (partials: {
  top: string[];
  head: string[];
  body: string[];
}) => `
<!DOCTYPE html>
<html>
<head>

  ${partials.top.join('\n')}

  ${partials.head.join('\n')}

</head>

<body>
  <div id="<%= mountId %>"><!--<?- html ?>--></div>
  ${partials.body.join('\n')}
  <!--<?- chunksMap.js ?>-->
  <!--<?- SSRDataScript ?>-->
  <!--<?- bottomTemplate ?>-->
</body>

</html>
`;
