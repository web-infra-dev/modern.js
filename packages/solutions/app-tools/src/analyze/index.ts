import * as path from 'path';
import { createDebugger, fs, isApiOnly } from '@modern-js/utils';
import type { CliPlugin } from '@modern-js/core';
import { cloneDeep } from '@modern-js/utils/lodash';
import { createBuilderForModern } from '../builder';
import { printInstructions } from '../utils/printInstructions';
import { generateRoutes } from '../utils/routes';
import { emitResolvedConfig } from '../utils/config';
import { getCommand } from '../utils/commands';
import { AppTools } from '../types';
import { initialNormalizedConfig } from '../config';
import { isRouteComponentFile } from './utils';
import { loaderBuilder, serverLoaderBuilder } from './Builder';

const debug = createDebugger('plugin-analyze');

export default (): CliPlugin<AppTools> => ({
  name: '@modern-js/plugin-analyze',

  setup: api => {
    let pagesDir: string[] = [];
    let originEntrypoints: any[] = [];

    return {
      async prepare() {
        let appContext = api.useAppContext();
        const resolvedConfig = api.useResolvedConfigContext();
        const hookRunners = api.useHookRunners();

        try {
          fs.emptydirSync(appContext.internalDirectory);
        } catch {
          // FIXME:
        }

        const apiOnly = await isApiOnly(
          appContext.appDirectory,
          resolvedConfig.source?.entriesDir,
        );
        await hookRunners.addRuntimeExports();

        if (apiOnly) {
          const { routes } = await hookRunners.modifyServerRoutes({
            routes: [],
          });

          debug(`server routes: %o`, routes);

          appContext = {
            ...appContext,
            apiOnly,
            serverRoutes: routes,
          };
          api.setAppContext(appContext);
          return;
        }

        const [
          { getBundleEntry },
          { getServerRoutes },
          { generateCode },
          { getHtmlTemplate },
        ] = await Promise.all([
          import('./getBundleEntry'),
          import('./getServerRoutes'),
          import('./generateCode'),
          import('./getHtmlTemplate'),
        ]);

        const entrypoints = getBundleEntry(appContext, resolvedConfig);
        const defaultChecked = entrypoints.map(point => point.entryName);

        debug(`entrypoints: %o`, entrypoints);

        const initialRoutes = getServerRoutes(entrypoints, {
          appContext,
          config: resolvedConfig,
        });

        const { routes } = await hookRunners.modifyServerRoutes({
          routes: initialRoutes,
        });

        debug(`server routes: %o`, routes);

        appContext = {
          ...appContext,
          entrypoints,
          serverRoutes: routes,
        };
        api.setAppContext(appContext);

        const nestedRouteEntries = entrypoints
          .map(point => point.nestedRoutesEntry)
          .filter(Boolean) as string[];

        pagesDir = entrypoints
          .map(point => point.entry)
          .filter(Boolean)
          .concat(nestedRouteEntries);

        originEntrypoints = cloneDeep(entrypoints);

        await generateCode(appContext, resolvedConfig, entrypoints, api);

        const htmlTemplates = await getHtmlTemplate(entrypoints, api, {
          appContext,
          config: resolvedConfig,
        });

        debug(`html templates: %o`, htmlTemplates);

        await hookRunners.addDefineTypes();

        debug(`add Define Types`);

        appContext = {
          ...appContext,
          entrypoints,
          checkedEntries: defaultChecked,
          apiOnly,
          serverRoutes: routes,
          htmlTemplates,
        };
        api.setAppContext(appContext);

        const command = getCommand();
        const buildCommands = ['dev', 'build', 'inspect', 'deploy'];
        if (buildCommands.includes(command)) {
          const normalizedConfig = api.useResolvedConfigContext();
          const builder = await createBuilderForModern({
            normalizedConfig: normalizedConfig as any,
            appContext,
            compatPluginConfig: {
              async onBeforeBuild({ bundlerConfigs }) {
                const hookRunners = api.useHookRunners();
                await generateRoutes(appContext);
                await hookRunners.beforeBuild({ bundlerConfigs });
              },

              async onAfterBuild({ stats }) {
                const hookRunners = api.useHookRunners();
                await hookRunners.afterBuild({ stats });
                await emitResolvedConfig(
                  appContext.appDirectory,
                  normalizedConfig,
                );
              },

              async onDevCompileDone({ isFirstCompile }) {
                const hookRunners = api.useHookRunners();
                if (process.stdout.isTTY || isFirstCompile) {
                  hookRunners.afterDev();

                  if (isFirstCompile) {
                    printInstructions(
                      hookRunners,
                      appContext,
                      normalizedConfig,
                    );
                  }
                }
              },

              async onBeforeCreateCompiler({ bundlerConfigs }) {
                const hookRunners = api.useHookRunners();
                // run modernjs framework `beforeCreateCompiler` hook
                await hookRunners.beforeCreateCompiler({
                  bundlerConfigs,
                });
              },

              async onAfterCreateCompiler({ compiler }) {
                const hookRunners = api.useHookRunners();
                // run modernjs framework afterCreateCompiler hooks
                await hookRunners.afterCreateCompiler({ compiler });
              },
            },
          });

          builder.addPlugins(resolvedConfig.builderPlugins);

          appContext = {
            ...appContext,
            builder,
          };
          api.setAppContext(appContext);
        }
      },
      watchFiles() {
        return pagesDir;
      },

      resolvedConfig({ resolved }) {
        const appContext = api.useAppContext();
        const config = initialNormalizedConfig(resolved, appContext);
        return {
          resolved: config,
        };
      },

      async beforeRestart() {
        serverLoaderBuilder.stop();
        loaderBuilder.stop();
      },

      async fileChange(e) {
        const appContext = api.useAppContext();
        const { appDirectory } = appContext;
        const { filename, eventType } = e;
        const isPageFile = (name: string) =>
          pagesDir.some(pageDir => name.includes(pageDir));

        const absoluteFilePath = path.resolve(appDirectory, filename);
        const isRouteComponent =
          isPageFile(absoluteFilePath) &&
          isRouteComponentFile(absoluteFilePath);

        if (
          isRouteComponent &&
          (eventType === 'add' || eventType === 'unlink')
        ) {
          const resolvedConfig = api.useResolvedConfigContext();
          const { generateCode } = await import('./generateCode');
          const entrypoints = cloneDeep(originEntrypoints);
          generateCode(appContext, resolvedConfig, entrypoints, api);
        }
      },
    };
  },
});
