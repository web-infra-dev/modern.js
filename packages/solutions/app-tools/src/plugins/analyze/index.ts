import * as path from 'path';
import {
  createDebugger,
  findExists,
  fs,
  isApiOnly,
  minimist,
  isDevCommand,
  getArgv,
} from '@modern-js/utils';
import type { CliPlugin } from '@modern-js/core';
import { printInstructions } from '../../utils/printInstructions';
import { generateRoutes, getPathWithoutExt } from '../../utils/routes';
import { emitResolvedConfig } from '../../utils/config';
import { getSelectedEntries } from '../../utils/getSelectedEntries';
import { AppTools, webpack } from '../../types';
import { initialNormalizedConfig } from '../../config';
import { createBuilderGenerator } from '../../builder';
import { checkIsBuildCommands, parseModule, replaceWithAlias } from './utils';
import {
  APP_CONFIG_NAME,
  APP_INIT_EXPORTED,
  APP_INIT_IMPORTED,
} from './constants';
import { generateIndexCode } from './generateCode';

const debug = createDebugger('plugin-analyze');

export default ({
  bundler,
}: {
  bundler: 'webpack' | 'rspack';
}): CliPlugin<AppTools<'shared'>> => ({
  name: '@modern-js/plugin-analyze',

  setup: api => {
    let pagesDir: string[] = [];
    let nestedRouteEntries: string[] = [];

    return {
      async prepare() {
        let appContext = api.useAppContext();
        const resolvedConfig = api.useResolvedConfigContext();
        const hookRunners = api.useHookRunners();

        try {
          if (checkIsBuildCommands()) {
            fs.emptydirSync(appContext.internalDirectory);
          }
        } catch {
          // FIXME:
        }

        const apiOnly = await isApiOnly(
          appContext.appDirectory,
          resolvedConfig.source?.entriesDir,
          appContext.apiDirectory,
        );
        await hookRunners.addRuntimeExports();

        if (apiOnly) {
          const { routes } = await hookRunners.modifyServerRoutes({
            routes: [],
          });

          debug(`server routes: %o`, routes);

          appContext = {
            ...api.useAppContext(),
            apiOnly,
            serverRoutes: routes,
          };
          api.setAppContext(appContext);
          return;
        }

        const [
          { getBundleEntry },
          { getServerRoutes },
          { generateCode },
          { getHtmlTemplate },
        ] = await Promise.all([
          import('./getBundleEntry'),
          import('./getServerRoutes'),
          import('./generateCode'),
          import('./getHtmlTemplate'),
        ]);

        // get runtime entry points
        const { entrypoints } = await hookRunners.modifyEntrypoints({
          entrypoints: await getBundleEntry(
            hookRunners,
            appContext,
            resolvedConfig,
          ),
        });

        debug(`entrypoints: %o`, entrypoints);

        const initialRoutes = getServerRoutes(entrypoints, {
          appContext,
          config: resolvedConfig,
        });

        const { routes } = await hookRunners.modifyServerRoutes({
          routes: initialRoutes,
        });

        debug(`server routes: %o`, routes);

        appContext = {
          ...api.useAppContext(),
          entrypoints,
          serverRoutes: routes,
        };
        api.setAppContext(appContext);

        nestedRouteEntries = entrypoints
          .map(point => point.nestedRoutesEntry)
          .filter(Boolean) as string[];

        pagesDir = entrypoints
          .map(point => point.entry)
          // should only watch file-based routes
          .filter(entry => entry && !path.extname(entry))
          .concat(nestedRouteEntries);

        const { importsStatemets } = await generateCode(
          appContext,
          resolvedConfig,
          entrypoints,
          api,
        );

        const htmlTemplates = await getHtmlTemplate(entrypoints, api, {
          appContext,
          config: resolvedConfig,
        });

        debug(`html templates: %o`, htmlTemplates);

        await hookRunners.addDefineTypes();

        debug(`add Define Types`);

        let checkedEntries = entrypoints.map(point => point.entryName);
        if (isDevCommand()) {
          const { entry } = minimist(getArgv());
          checkedEntries = await getSelectedEntries(
            typeof entry === 'string' ? entry.split(',') : entry,
            entrypoints,
          );
        }

        appContext = {
          ...api.useAppContext(),
          entrypoints,
          checkedEntries,
          apiOnly,
          serverRoutes: routes,
          htmlTemplates,
        };

        api.setAppContext(appContext);

        if (checkIsBuildCommands()) {
          const normalizedConfig = api.useResolvedConfigContext();
          const createBuilderForModern = await createBuilderGenerator(bundler);
          const builder = await createBuilderForModern({
            normalizedConfig: normalizedConfig as any,
            appContext,
          });

          builder.onBeforeBuild(async ({ bundlerConfigs }) => {
            const hookRunners = api.useHookRunners();
            await generateRoutes(appContext);
            await hookRunners.beforeBuild({
              bundlerConfigs:
                bundlerConfigs as unknown as webpack.Configuration[],
            });
          });

          builder.onAfterBuild(async ({ stats }) => {
            const hookRunners = api.useHookRunners();
            await hookRunners.afterBuild({ stats });
            await emitResolvedConfig(appContext.appDirectory, normalizedConfig);
          });

          builder.onDevCompileDone(async ({ isFirstCompile }) => {
            const hookRunners = api.useHookRunners();
            hookRunners.afterDev({ isFirstCompile });

            if (isFirstCompile) {
              printInstructions(hookRunners, appContext, normalizedConfig);
            }
          });

          builder.onBeforeCreateCompiler(async ({ bundlerConfigs }) => {
            const hookRunners = api.useHookRunners();
            await generateIndexCode({
              appContext,
              config: resolvedConfig,
              entrypoints,
              api,
              importsStatemets,
              bundlerConfigs,
            });

            // run modernjs framework `beforeCreateCompiler` hook
            await hookRunners.beforeCreateCompiler({
              bundlerConfigs:
                bundlerConfigs as unknown as webpack.Configuration[],
            });
          });

          builder.onAfterCreateCompiler(async ({ compiler }) => {
            const hookRunners = api.useHookRunners();
            // run modernjs framework afterCreateCompiler hooks
            await hookRunners.afterCreateCompiler({
              compiler: compiler as unknown as
                | webpack.Compiler
                | webpack.MultiCompiler,
            });
          });

          builder.addPlugins(resolvedConfig.builderPlugins);

          appContext = {
            ...api.useAppContext(),
            builder,
          };
          api.setAppContext(appContext);
        }
      },

      watchFiles() {
        return { files: pagesDir, isPrivate: true };
      },

      resolvedConfig({ resolved }) {
        const appContext = api.useAppContext();
        const config = initialNormalizedConfig(resolved, appContext, bundler);
        return {
          resolved: config,
        };
      },

      // This logic is not in the router plugin to avoid having to include some dependencies in the utils package
      async modifyEntryImports({ entrypoint, imports }) {
        const appContext = api.useAppContext();
        const { srcDirectory, internalSrcAlias } = appContext;
        const { fileSystemRoutes, nestedRoutesEntry } = entrypoint;
        if (fileSystemRoutes && nestedRoutesEntry) {
          const rootLayoutPath = path.join(nestedRoutesEntry, 'layout');
          const rootLayoutFile = findExists(
            ['.js', '.ts', '.jsx', '.tsx'].map(
              ext => `${rootLayoutPath}${ext}`,
            ),
          );
          if (rootLayoutFile) {
            const rootLayoutBuffer = await fs.readFile(rootLayoutFile);
            const rootLayout = rootLayoutBuffer.toString();
            const [, moduleExports] = await parseModule({
              source: rootLayout.toString(),
              filename: rootLayoutFile,
            });
            const hasAppConfig = moduleExports.some(
              e => e.n === APP_CONFIG_NAME,
            );
            const generateLayoutPath = getPathWithoutExt(
              replaceWithAlias(srcDirectory, rootLayoutFile, internalSrcAlias),
            );

            if (hasAppConfig) {
              imports.push({
                value: generateLayoutPath,
                specifiers: [{ imported: APP_CONFIG_NAME }],
              });
            }

            const hasAppInit = moduleExports.some(
              e => e.n === APP_INIT_EXPORTED,
            );

            if (hasAppInit) {
              imports.push({
                value: generateLayoutPath,
                specifiers: [
                  { imported: APP_INIT_EXPORTED, local: APP_INIT_IMPORTED },
                ],
              });
            }
          }
        }
        return {
          entrypoint,
          imports,
        };
      },
    };
  },
});
