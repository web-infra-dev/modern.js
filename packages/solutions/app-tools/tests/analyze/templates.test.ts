/* eslint-disable @typescript-eslint/no-empty-function */
import { renderFunction } from '../../src/plugins/analyze/templates';

jest.mock('@modern-js/utils', () => {
  const fs = {
    writeFile() {},
    writeJSON() {},
    ensureFile() {},
  };
  return {
    __esModule: true,
    ...jest.requireActual('@modern-js/utils'),
    fs,
  };
});

expect.addSnapshotSerializer({
  test: val => typeof val === 'string',
  print: (val: unknown) => (val as string).replace(/\\/g, '/'),
});

describe('renderFunction', () => {
  test('basic usage', () => {
    const code = renderFunction({
      plugins: [],
      customBootstrap: false,
      fileSystemRoutes: undefined,
    });

    expect(code).toMatchSnapshot();
  });
});
