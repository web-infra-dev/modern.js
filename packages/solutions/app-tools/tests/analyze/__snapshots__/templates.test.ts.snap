// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderFunction basic usage 1`] = `

  const finalAppConfig = {
    ...App.config,
    ...typeof config === 'function' ? config() : {},
  }

  AppWrapper = createApp({
    plugins: [
     
       
    ]
  })(App)


  if(!AppWrapper.init && typeof appInit !== 'undefined') {
    AppWrapper.init = appInit;
  }


  if (IS_BROWSER) {
    bootstrap(AppWrapper, MOUNT_ID, root, ReactDOM);
  }

  return AppWrapper

`;
